airflow/models/taskinstance.py

```diff
@@ -1632,11 +1632,12 @@ def _get_previous_ti(
 
 @internal_api_call
 @provide_session
-def _update_rtif(ti, rendered_fields, session: Session | None = None):
+def _update_rtif(ti, rendered_fields, session: Session = NEW_SESSION):
     from airflow.models.renderedtifields import RenderedTaskInstanceFields
 
     rtif = RenderedTaskInstanceFields(ti=ti, render_templates=False, rendered_fields=rendered_fields)
     RenderedTaskInstanceFields.write(rtif, session=session)
+    session.commit()
```

Comment:
I don't have a strong opinion on this, but if my memory serves me right, the last time @uranusjr mentioned we should prefer `session.flush()` over `session.commit()` when possible, but I'm not sure whether this is possible here 🤔

Reply:
yeah i think flush actually might work here.  if we have a test that repros the bug we could verify. not sure if we can produce such a test.  it is very odd that it would be deleting the record it's trying to update.

Reply:
ok so i think i understand what's going on

RTIF pk is same as TI

So RTIF to TI is 1-1

but there is logic to keep only the "latetst" N RTIF records, controlled by `max_num_rendered_ti_fields_per_task`

so if we're already at the limit of RTIF records, and we add one more, it will always be deleted.

alternatively we could run the delete code first and _then_ write.  that would probably make more sense.

Reply:
Delete first would need to delete if current count is n-1, else we would have n+1 after deleting then writing a new one.

Reply:
>I don't have a strong opinion on this, but if my memory serves me right, the last time @uranusjr mentioned we should prefer session.flush() over session.commit() when possible, but I'm not sure whether this is possible here 🤔

It looks like `session.commit()` calls `session.flush()`: https://docs.sqlalchemy.org/en/20/orm/session_basics.html#flushing and flush is also called before any individual sql statement/query